### ISSUE TITLE
Null-coalescing operator with nullable reference types gives incorrect warning

### ISSUE THREAD

**Issue by joukokynsijarvi** [dotnet/roslyn#42896]
**Version Used**: csc 3.6.0-2.20173.3 (2242fa4f) / VS 16.6.0 Preview 2.0

**Steps to Reproduce**:

Create a .NET Core 3.1 console project with following code:

```c#
using System;

class Program
{
    static void Main()
    {
        string? foo = null;
        var bar = foo ?? String.Empty;
        var baz = bar ?? String.Empty;

        var b = bar.Length;
        var z = baz.Length;
    }
}
```

**Expected Behavior**:

"var b = bar.Length" should not give warning CS8602  Dereference of a possibly null reference

**Actual Behavior**:

"var b = bar.Length" gives warning CS8602  Dereference of a possibly null reference

**Comment by YairHalberstadt** [comment]
I think whats going on here, is that `bar ?? String.Empty` is taken as you explicitly stating that bar can be null. So when you later check `bar.Length` without assigning to it it warns.

See https://github.com/dotnet/csharplang/blob/f231090bb17109f95d2d699f4f5db4c7a06b495a/meetings/2019/LDM-2019-02-20.md

**Comment by jcouv** [comment]
@YairHalberstadt is correct.

Certain tests on a variable, such as `is null` or `== null` or `??` or `?.`, tell the compiler you meant the variable has a maybe-null state at that point.
In this case, `bar ?? String.Empty` changes the state of `bar` to maybe-null, thus the following warning.
One rationale for this design is so that you can write null tests in public APIs that declare non-nullable inputs.

This is by-design.
